crates/storage/src/lazy/mod.rs

```diff
@@ -15,7 +15,252 @@
+pub use self::mapping::Mapping;
+
+use crate::traits::{
+    pull_storage,
+    push_storage,
+    AutoKey,
+    DecodeWrapper,
+    Item,
+    KeyHolder,
+    Storable,
+};
+use core::marker::PhantomData;
+use ink_primitives::Key;
+use scale::{
+    Error,
+    Input,
+    Output,
+};
+
+/// A simple wrapper around type to store it in a separate storage cell under own storage key.
+/// If you want to update the value, first you need to `get` it, update, and after `set`.
+///
+/// # Important
+///
+/// The wrapper requires its own pre-defined storage key where to store value. By default,
+/// it is [`AutoKey`](crate::traits::AutoKey) and during compilation is calculated based on
+/// the name of the structure and the field. But anyone can specify its storage key
+/// via [`ManualKey`](crate::traits::ManualKey).
+///
+/// This is an example of how you can do this:
+/// ```rust
+/// # use ink_lang as ink;
+/// # use ink_env::{
+/// #     Environment,
+/// #     DefaultEnvironment,
+/// # };
+/// # type AccountId = <DefaultEnvironment as Environment>::AccountId;
+///
+/// # #[ink::contract]
+/// # mod my_module {
+/// use ink_storage::{traits::ManualKey, Lazy};
+///
+/// #[ink(storage)]
+/// #[derive(Default)]
+/// pub struct MyContract {
+///     owner: Lazy<AccountId>,
+///     balance: Lazy<Balance, ManualKey<123>>,
+/// }
+///
+/// impl MyContract {
+///     #[ink(constructor)]
+///     pub fn new() -> Self {
+///         let mut instance = Self::default();
+///         instance.new_init();
+///         instance
+///     }
+///
+///     /// Default initializes the contract.
+///     fn new_init(&mut self) {
+///         let caller = Self::env().caller();
+///         self.owner.set(&caller);
+///         self.balance.set(&123456);
+///     }
+///
+/// #   #[ink(message)]
+/// #   pub fn my_message(&self) { }
+/// }
+/// # }
+/// ```
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+pub struct Lazy<V, KeyType: KeyHolder = AutoKey> {
+    _marker: PhantomData<fn() -> (V, KeyType)>,
+}
+
+/// We implement this manually because the derived implementation adds trait bounds.
+impl<V, KeyType> Default for Lazy<V, KeyType>
+where
+    KeyType: KeyHolder,
+{
+    fn default() -> Self {
+        Self {
+            _marker: Default::default(),
+        }
+    }
+}
+
+impl<V, KeyType> Lazy<V, KeyType>
+where
+    KeyType: KeyHolder,
+{
+    /// Creates a new empty `Lazy`.
+    pub fn new() -> Self {
+        Self {
+            _marker: Default::default(),
+        }
+    }
+}
+
+impl<V, KeyType> core::fmt::Debug for Lazy<V, KeyType>
+where
+    KeyType: KeyHolder,
+{
+    fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+        f.debug_struct("Lazy").field("key", &KeyType::KEY).finish()
+    }
+}
+
+impl<V, KeyType> Lazy<V, KeyType>
+where
+    V: Storable,
+    KeyType: KeyHolder,
+{
+    /// Get the `value` from the contract storage.
+    ///
+    /// Panics if no `value` exists.
+    pub fn get(&self) -> V {
```

Comment:
So `Mapping::get()` returns an `Option`, maybe we should do that instead of panicking. We can think about adding a `Deref` impl or something if people want to be more 🤠 about accessing the inner contents